nncf/common/pruning/utils.py

```diff
@@ -0,0 +1,214 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+import math
+from functools import partial
+from typing import List, Tuple, Optional
+
+from nncf.common.graph.graph import NNCFGraph
+from nncf.common.graph.graph import NNCFNode
+from nncf.common.graph.module_attributes import ConvolutionModuleAttributes
+from nncf.common.utils.registry import Registry
+
+
+def is_grouped_conv(node: NNCFNode) -> bool:
+    return isinstance(node.module_attributes, ConvolutionModuleAttributes) \
+           and node.module_attributes.groups != 1
+
+
+def get_sources_of_node(nncf_node: NNCFNode, graph: NNCFGraph, sources_types: List[str]) -> List[NNCFNode]:
+    """
+    Source is a node of source such that there is path from this node to nncf_node and on this path
+    no node has one of sources_types type.
+    :param sources_types: list of sources types
+    :param nncf_node: NNCFNode to get sources
+    :param graph: NNCF graph to work with
+    :return: list of all sources nodes
+    """
+    visited = {node_id: False for node_id in graph.get_all_node_idxs()}
+    partial_traverse_function = partial(traverse_function, type_check_fn=lambda x: x in sources_types,
+                                        visited=visited)
+    nncf_nodes = [nncf_node]
+    if nncf_node.node_type in sources_types:
+        nncf_nodes = graph.get_previous_nodes(nncf_node)
+
+    source_nodes = []
+    for node in nncf_nodes:
+        source_nodes.extend(graph.traverse_graph(node, partial_traverse_function, False))
+    return source_nodes
```

Comment:
Should be moved to NNCFGraph. Can this be done in this PR?

Reply:
There is no difficulties in doing this. I'm just not sure whether these methods are general enough to be in NNCFGraph.
@alexsu52 please comment on this and following suggestions about moving methods to NNCFGraph, since NNCFGraph is a common part.

Reply:
It looks like that these methods are used only in the pruning algorithm and implements very specific algorithms. Probably we should refactor this part of the code and create a module with common algorithms which are worked with the NNCFGraph. 

@vshampor could you provide justification why it should be moved to NNCFGraph?

Reply:
1. The file name is `utils.py`, which in most cases means that all of the functions there can be moved to one class or the other.
2. Most of these functions cannot function without NNCFGraph as its first argument and will only function if NNCFGraph is passed as its first argument. I see transition of these functions to NNCFGraph and replacing `graph` with `self` as a quite natural refactoring. At the very least, if all of these functions only use NNCFGraph's public methods and you want to keep the type of analysis performed by these functions local to the pruning algo - then create a correspondingly named pruning-specific analyzer class (if it is not yet created), initialize it with an NNCFGraph instance and move these functions to become methods of this analyzer class.

Reply:
1. I still doubt that these methods should be in NNCFGraph
2. I agree that it would be good to reduce using of utils, but currently in NNCF there are different utils, furthermore there are utils in original NNCF pruning. Goal of this PR is generalization of filter pruning algorithm, not its refactoring. I've created epic for NNCF utils refactoring: 51044.